Balance pass and bug fixes for scripts

z9484 opened the pull request.

z9484 committed:
Balance pass and bug fixes for scripts

OnlineCop commented:
Sorry for the delay; I've been fighting off another round of Covid.

There are a lot of changes here; could you describe what kinds of balancing we're looking at here?

z9484 committed:
Merge branch 'road_to_1.0' into balancePass

z9484 committed:
Fixed set ent active and facing for cave8; Added KQ_SAVEDIR to change save location for different kq modules; Fixed Portbubble height

z9484 commented:
So I played KQ from the beginning to the end without cheating (mostly). I wanted to make sure the mobs didn't have any crazy difficulty spikes that would require extensive grinding. But I really only changed the stats for the new ones I had added and didn't mess with the originals. I also had to fix some script errors. 

I added in a fix for set_ent_active which had gotten broken recently. I also added a new CMAKE variable that will change save folder location to better support different games with the KQ engine.

OnlineCop commented:
Is it possible to move the cmake changes to their own pull request?

It appears that the draw.cpp changes are a separate issue, and the intrface.cpp changes are a different one still.

OnlineCop commented on src/draw.cpp:
```diff
@@ -710,11 +710,11 @@ void KDraw::draw_porttextbox(eBubbleStyle bstyle, int chr)
 
-    menubox(double_buffer, 19, 172 - linexofs, 4, 4, BLUE);
-    menubox(double_buffer, 66, 196 - linexofs, party[chr].name.length(), 1, BLUE);
+    menubox(double_buffer, 19, 154 - linexofs, 4, 4, BLUE);
```
Do we have any idea what these magic numbers are? `172` isn't evenly divisible by 16 (TILE_W), nor is `154` (which is now 18 less than 172).

If they're offsets from the left edge of, say, a border/frame, then a comment or `const int FrameXoffset = ...;` would be useful to see what it implies.

If it's just "it looks good to my eyeball", that's okay, but I'd like to remove as many "magic numbers" that I find in the code as I can.

z9484 replied:
Yes its just what looks good. This function needs some cleanup as why is a line X offset being subtracted from the Y coordinate?

OnlineCop commented on scripts/cave8.lua:
```diff
@@ -97,11 +97,13 @@ function entity_handler(en)
         x, y = marker("nostik_e")
         move_entity(1, x, y, 0)
-        wait_for_entity(1, 2)
+        wait_for_entity(HERO1, 1)
```
This commit looks like it should be the `cmake` stuff. Could you move these changes to a separate commit? (It doesn't have to be moved to a new pull request.)

OnlineCop commented on src/intrface.cpp:
```diff
@@ -2960,7 +2960,7 @@ static int KQ_set_desc(lua_State* L)
 {
     int a = real_entity_num(L, 1);
-    auto b = lua_toboolean(L, 2);
+    int b = lua_tointeger(L, 2);
```
Was `lua_toboolean()` the cause that entities weren't being set active/inactive?

z9484 replied:
Looks like it was converting "0" to 1 for some reason

OnlineCop commented on src/unix.cpp:
```diff
@@ -149,7 +149,11 @@ const std::string kqres(enum eDirectories dir, const std::string& file)
-        user_dir = std::string(SDL_GetPrefPath("kq-fork", "kq"));
+        std::string save_folder = "kq";
+#ifdef KQ_SAVEDIR
+        save_folder = std::string { KQ_SAVEDIR };
```
If `KQ_SAVEDIR` is defined BUT empty, what will happen?

OnlineCop commented on scripts/grotto2.lua:
```diff
@@ -25,37 +26,30 @@ function entity_handler(en)
   elseif (get_ent_id(en) == CASANDRA) then
-    -- bubble(en, _"I can use my Boost ability to strengthen spells when I am attacking.")
-    bubble(en, _"It's time.")
+    bubble(en, _"Malkaron won't know what hits him.")
```
NIT: ...won't know what `hit` him...

OnlineCop commented on scripts/tunnela.lua:
```diff
@@ -42,5 +49,5 @@ function zone_handler(zn)
   --   change_map("tunnela", "entrance_w")
-  end
+  -- end
 end
```
NIT: Needs a newline at the bottom of the file.

OnlineCop reviewed (approved):
Are data/fortress.png and maps/fortress.png the same?

z9484 commented:
Yes images are stored in data and in maps. Looks like we can create an issue to delete all the tileset images in data because I don't think they are actually being used.
